File: flyweight/typescript/box-factory.ts
import {BoxColor} from "./box-color";
import {Box} from "./box";
import {BlackBox} from "./black-box";
import {BlueBox} from "./blue-box";

// Flyweight (Box) havuzunu tutacak sınıftır.
// Eğer daha önceden nesne oluşturulmamışsa havuza atılacak.
// Oluşturulmuş ise de havuzdan seçilip verilecektir.
// UML diyagramındaki FlyweightFactory sınıfına denk gelmektedir.
export class BoxFactory {
  private readonly boxes: Map<BoxColor, Box>;

  constructor() {
    this.boxes = new Map<BoxColor, Box>();
  }

  public get(boxColor: BoxColor): Box {
    if(this.boxes.has(boxColor)) return this.boxes[boxColor];

    let box: Box;

    switch (boxColor) {
      case BoxColor.BLACK:
        box = new BlackBox(20, 20);
        break;

      case BoxColor.BLUE:
        box = new BlueBox(20, 20);
        break;

      default:
        throw new Error('Unsupported box color!');
    }

    this.boxes.set(boxColor, box);
    return box;
  }
}

File: flyweight/go/box-factory.go
package main

// Flyweight (Box) havuzunu tutacak sınıftır.
// Eğer daha önceden nesne oluşturulmamışsa havuza atılacak.
// Oluşturulmuş ise de havuzdan seçilip verilecektir.
// UML diyagramındaki FlyweightFactory sınıfına denk gelmektedir.
type boxFactory struct {
	boxes map[int]Box
}

func NewBoxFactory() *boxFactory {
	return &boxFactory{
		boxes: make(map[int]Box),
	}
}

func (f *boxFactory) Get(boxColor int) Box {
	if box, ok := f.boxes[boxColor]; ok {
		return box
	}

	var box Box

	switch boxColor {
	case BLACK:
		box = NewBlackBox(20, 20)
	case BLUE:
		box = NewBlueBox(20, 20)
	}

	f.boxes[boxColor] = box
	return box
}
